File: java/CAchievement.java
package cell;

import java.util.Vector;

class CQuest implements ISerialize
{
	int m_offset;
	String m_img;
	CVPAndOr m_success=new CVPAndOr();
	String m_location;
	boolean m_guide;
	boolean m_cameraMove;
	CCommunityAction m_cmt=new CCommunityAction();
	public CPacket Serialize()
	{
		CPacket pac=new CPacket();
		pac.name="CQuest";
		pac.Push(m_offset);
		pac.Push(m_img);
		pac.Push(m_success);
		pac.Push(m_location);
		pac.Push(m_guide);
		pac.Push(m_cameraMove);
		pac.Push(m_cmt);

		return pac;
	}
	public void Deserialize(String _str) 
	{
		CPacket pac=new CPacket();
		pac.Deserialize(_str);
		m_offset=Integer.parseInt(pac.value.get(0));
		m_img=pac.value.get(1);
		m_success.Deserialize(pac.value.get(2));
		m_location=pac.value.get(3);
		m_guide=pac.value.get(4).equals("true");
		m_cameraMove=pac.value.get(5).equals("true");
		m_cmt.Deserialize(pac.value.get(6));
	}
}
public class CAchievement implements ISerialize
{
	int m_offset;
	String m_subject;
	String m_img;
	CVPAndOr m_revelation=new CVPAndOr();//¹ßÇö
	String m_zone;
	Vector<CQuest> m_quest=new Vector<CQuest>();
	Vector<CDropInfo> m_item=new Vector<CDropInfo>();
	
	public CPacket Serialize()
	{
		CPacket pac=new CPacket();
		pac.name="CAchievement";
		pac.Push(m_offset);
		pac.Push(m_subject);
		pac.Push(m_img);
		pac.Push(m_revelation);
		pac.Push(m_zone);
		pac.Push(m_quest);
		pac.Push(m_item);
		
		return pac;
	}
	public void Deserialize(String _str) 
	{
		CPacket pac=new CPacket();
		pac.Deserialize(_str);
		m_offset=Integer.parseInt(pac.value.get(0));
		m_subject=pac.value.get(1);
		m_img=pac.value.get(2);
		m_revelation.Deserialize(pac.value.get(3));
		m_zone=pac.value.get(4);
		CPacket pac2=new CPacket();
		pac2.Deserialize(pac.value.get(5));
		for(String each0 : pac2.value)
		{
			CQuest quest=new CQuest();
			quest.Deserialize(each0);
			m_quest.add(quest);
		}
		CPacket pac3=new CPacket();
		pac3.Deserialize(pac.value.get(6));
		for(String each0 : pac3.value)
		{
			CDropInfo item=new CDropInfo();
			item.Deserialize(each0);
			m_item.add(item);
		}
	}
}
class CAcSc
{
	static public Vector<CAchievement> m_acVec=new Vector<CAchievement>();
	static public String m_script="";
	static public void Init()
	{
		
		if(m_script.isEmpty())
			m_script=CDbMgr.GetText("Ac");
		if(m_script.isEmpty())
			return;
		CPacket pac=new CPacket();
		pac.Deserialize(m_script);
		for(String each0 : pac.value)
		{
			CAchievement ac=new CAchievement();
			ac.Deserialize(each0);
			m_acVec.add(ac);
		}
	}
	static public class CAQInfo
	{
		CAQInfo(int _a,int _q)
		{
			a=_a;
			q=_q;
		}
		int a=-1;
		int q=-1;
	}
	static public CAQInfo Chk(CUser _user)
	{
		if(_user.GetAchievement())
			return null;
		
		for(CAchievement each0 : m_acVec)
		{
			int count=_user.GetVMapValue(VK.Get(Df.VMap.Qt,each0.m_offset));
			if(each0.m_revelation.Process(_user.m_vMap) ||
					count>0)
			{
				
				if(each0.m_quest.size()<count)
				{
					continue;
				}
				else if(each0.m_quest.size()==count)
				{
					_user.VMapPushPop(VK.Get(Df.VMap.Qt,each0.m_offset), each0.m_quest.size()+1);
					for(CDropInfo each1 : each0.m_item)
					{
						if(each1.m_amount<0 && _user.InvenInChk(each1.m_itemOff, -each1.m_amount))
						{
							_user.SetInvenOffset(_user.FindItemtoOffset(each1.m_itemOff), each1.m_amount);
						}
						else
							_user.PushInven(new CInventoryInfo(each1.m_itemOff, each1.m_amount));
					}
				}
				else if(each0.m_quest.size()>count)
				{
					if(each0.m_quest.get(count).m_success.Process(_user.m_vMap)==true)
					{
						if(each0.m_quest.get(count).m_cmt.m_text.isEmpty())
						{
							_user.VMapPushPop(VK.Get(Df.VMap.Qt,each0.m_offset), count+1);
							continue;
						}
						
						return new CAQInfo(each0.m_offset,count);
					}
				}
	
				
			}
		}
		return null;
	}
}

File: js/CAchievement.js
function CQuest()
{
	this.m_offset;
	this.m_img;
	this.m_success=new CVPAndOr();
	this.m_location;
	this.m_guide;
	this.m_cameraMove;
	this.m_cmt=new CCommunityAction();
	
	
}
CQuest.prototype=new ISerialize();
CQuest.prototype.constructor=CQuest;
CQuest.prototype.Serialize=function()
{
	var pac=new CPacket();
	pac.name="CQuest";
	pac.Push(this.m_offset);
	pac.Push(this.m_img);
	pac.Push(this.m_success);
	pac.Push(this.m_location);
	pac.Push(this.m_guide);
	pac.Push(this.m_cameraMove);
	pac.Push(this.m_cmt);

	
	return pac;
}
CQuest.prototype.Deserialize=function(_str)
{
	var dummy=new CPacket();
	dummy.Deserialize(_str);
	
	this.m_offset=Number(dummy.value[0]);
	this.m_img=dummy.value[1];
	this.m_success.Deserialize(dummy.value[2]);
	this.m_location=dummy.value[3];
	this.m_guide=dummy.value[4]=="true";
	this.m_cameraMove=dummy.value[5]=="true";
	this.m_cmt.Deserialize(dummy.value[6]);
}
function CAchievement()
{
	this.m_offset;
	this.m_subject;
	this.m_img;
	this.m_revelation=new CVPAndOr();//CVProcess
	this.m_zone;
	this.m_quest=new Array();
	this.m_item=new Array();
}
CAchievement.prototype=new ISerialize();
CAchievement.prototype.constructor=CAchievement;
CAchievement.prototype.Serialize=function()
{
	var pac=new CPacket();
	pac.name="CAchievement";
	pac.Push(this.m_offset);
	pac.Push(this.m_subject);
	pac.Push(this.m_img);
	pac.Push(this.m_revelation);
	pac.Push(this.m_zone);
	pac.Push(this.m_quest);
	pac.Push(this.m_item);

	return pac;
}
CAchievement.prototype.Deserialize=function(_str)
{
	var dummy=new CPacket();
	dummy.Deserialize(_str);
	
	this.m_offset=Number(dummy.value[0]);
	this.m_subject=dummy.value[1];
	this.m_img=dummy.value[2];

	this.m_revelation.Deserialize(dummy.value[3]);
	this.m_zone=dummy.value[4];
	
	var dummy1=new CPacket();
	dummy1.Deserialize(dummy.value[5]);
	for(var i=0;i<dummy1.value.length;++i)
	{
		var quest=new CQuest();
		quest.Deserialize(dummy1.value[i]);
		this.m_quest.push(quest);
	}
	
	var dummy2=new CPacket();
	dummy2.Deserialize(dummy.value[6]);
	for(var i=0;i<dummy2.value.length;++i)
	{
		var item=new CDropInfo();
		item.Deserialize(dummy2.value[i]);
		this.m_item.push(item);
	}
}
var g_achArr=new Array();
